### PR TITLE
Update msfdb to support using pg_ctlcluster for db management

### PR DESCRIPTION
**Pull request by dwelch-r7**
First attempt at making use of the `pg_ctlcluster` utility for debian users

Still a little rough around the edges I'm sure a tonne more refactoring can be done

There are likely issues till to be resolved around some sort of upgrade path
One example I could see of something needing addressed is:
- Someone has a kali system set up to use the system postgres cluster
- They update to this script
- We will now attempt to use `pg_ctlcluster` to manage a new db rather than the system one they are currently using
- Their data is now all in the system cluster but they're connected to the new one

# Verification
- [ ] Running `./msfdb init` on ubuntu should now work out of the box (along with the rest of the commands: reinit, status, start stop, restart, delete)
- [ ] Installs on OSX should continue to work in the same way
- [ ] Check ubuntu versions as far back as 16.04
- [ ] We have a viable upgrade path for kali/black arch/other distros using the system cluster

Linking issues for visibility #12098 #11369 #14759

### CODE REVIEW

File: msfdb [hunk 1]
```diff
@@ -12,6 +12,10 @@ require 'rex/text'
 require 'securerandom'
 require 'uri'
 require 'yaml'
+require 'pg'
+require_relative 'msfdb_helpers/pg_ctlcluster'
```
Comment (adfoster-r7) [inline msfdb:16]: I wonder if these should live somewhere in `lib`? Adding a custom top level folder stuck out to me as being a smell 🤔

File: msfdb [hunk 2]
```diff
@@ -204,28 +187,30 @@ end
...
-    f.puts "port = #{@options[:db_port]}"
-  end
+def create_db_users
+  puts 'Creating database users'
+  run_psql("create user #{@options[:msf_db_user]} with password '#{@msf_pass}'")
+  run_psql("create user #{@options[:msftest_db_user]} with password '#{@msftest_pass}'")
+  run_psql("alter role #{@options[:msf_db_user]} createdb")
+  run_psql("alter role #{@options[:msftest_db_user]} createdb")
+  run_psql("alter role #{@options[:msf_db_user]} with password '#{@msf_pass}'")
+  run_psql("alter role #{@options[:msftest_db_user]} with password '#{@msftest_pass}'")
```
Comment (adfoster-r7) [inline msfdb:208]: Just confirming, should this interact with the driver object? 👀

File: msfdb [hunk 3]
```diff
@@ -204,28 +187,30 @@ end
...
+
+  conn = PG.connect(host: '127.0.0.1', dbname: 'postgres', port: @options[:db_port], user: @options[:msf_db_user], password: @msf_pass)
+  conn.exec("CREATE DATABASE #{@options[:msf_db_name]}")
+  conn.exec("CREATE DATABASE #{@options[:msftest_db_name]}")
+  conn.finish
```
Comment (adfoster-r7) [inline msfdb:213]: Any context on why these changes were needed in comparison to what was here before?

File: msfdb [hunk 4]
```diff
@@ -334,7 +309,7 @@ def write_db_config
 end
 
 def write_db_client_auth_config
```
Comment (adfoster-r7) [inline msfdb:311]: Is this dead code now?

File: msfdb [hunk 5]
```diff
@@ -1006,6 +968,10 @@ def invoke_command(commands, component, command)
   end
 end
 
+def installed(cmd)
```
Comment (adfoster-r7) [inline msfdb:971]: Predicate functions normally end with `?`, i.e. `installed?`

File: msfdb_helpers/db_interface.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
+class DbInterface
+
+  def init_db
```
Comment (adfoster-r7) [inline msfdb_helpers/db_interface.rb:3]: Should we nuke the `_db` suffix now that they're named already somewhat namespaced under a `db` class?

File: msfdb [hunk 7]
```diff
@@ -1006,6 +968,10 @@ def invoke_command(commands, component, command)
...
+  !Msf::Util::Helper.which(cmd).nil?
+end
+
 def has_requirements
```
Comment (adfoster-r7) [inline msfdb:975]: Is this functionality dropped now? 👀 

Looks like this had some good logging information when required tools were missing

**Review by adfoster-r7 — COMMENTED** [review]
I think this is a good step in the right direction :+1:

Is it worth exploring the docker effort as part of this pull request? It looks like it wouldn't take much more cycles to add now that the db logic is extracted into separate classes 📈

File: msfdb (hunk 1 above)
Comment (dwelch-r7) [reply msfdb:16]: I'm not against putting it in `lib` can do that for sure
is there something _wrong_ with putting things in their own top level folder though? just for future reference

File: msfdb (hunk 2 above)
Comment (dwelch-r7) [reply msfdb:208]: Nope! anything `psql` should be consistent across all types of database setups

File: msfdb (hunk 3 above)
Comment (dwelch-r7) [reply msfdb:213]: They aren't needed, worked fine the old way, but we have the `pg` gem anyway and I thought this was cleaner, I don't have overly strong feelings one way or the other though so happy to change it back if someone does

File: msfdb (hunk 4 above)
Comment (dwelch-r7) [reply msfdb:311]: sure is, good catch

File: msfdb (hunk 5 above)
Comment (dwelch-r7) [reply msfdb:971]: they sure do, will fix

File: msfdb_helpers/db_interface.rb (hunk 6 above)
Comment (dwelch-r7) [reply msfdb_helpers/db_interface.rb:3]: I actually started without the `_db` suffix and then added it in later :upside_down_face: 
I'm down to remove it

File: msfdb (hunk 7 above)
Comment (dwelch-r7) [reply msfdb:975]: I figured it was kind of superseded by `installed` and different DB setups need different thigns installed but I can probably add something similar tailored to each setup in without too much difficulty

**Comment by dwelch-r7** [comment]
> I think this is a good step in the right direction 
> 
> Is it worth exploring the docker effort as part of this pull request? It looks like it wouldn't take much more cycles to add now that the db logic is extracted into separate classes

I agree, I don't think it'll be too difficult, I'll have a crack at it

File: msfdb (hunk 7 above)
Comment (dwelch-r7) [reply msfdb:975]: I think I might be stupid :sweat_smile: 
it's still being used for the `pg_ctl` path but not `pg_ctlcluster` not sure why I did that, for sure can make it work for either

**Comment by adfoster-r7** [comment]
Thinking aloud, for the docker configuration potentially we could just support the database connection URIs format:
```
postgresql://[user[:password]@][netloc][:port][/dbname][?param1=value1&...]
```

https://www.postgresql.org/docs/10/libpq-connect.html#id-1.7.3.8.3.6

I haven't checked Ruby's PG bindings, but I'd guess there's parsing support there, or using URI.parse directly might be sufficient

**Comment by dwelch-r7** [comment]
@adfoster-r7 I'm pretty sure when I was messing around with the `pg` gem that it could deal with the URI, I'll keep it in mind

File: lib/msfdb_helpers/pg_ctlcluster.rb [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+  end
+
+  def self.requirements
+    %w(psql pg_ctlcluster pg_dropcluster pg_createcluster pg_config)
+  end
+
+  private
+
+  def get_postgres_version
+    _stdin, stdout, _stderr, _wait_thr = Open3.popen3('pg_config --version')
```
Comment (acammack-r7) [inline lib/msfdb_helpers/pg_ctlcluster.rb:107]:
```suggestion
    output, _status = Open3.capture2('pg_config --version')
```

File: lib/msfdb_helpers/pg_ctlcluster.rb [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+      exitstatus = wait_thr.value.exitstatus
+    end
+
+    if @options[:debug]
+      puts "'#{cmd}' returned #{exitstatus}"
+      puts out
+      puts err
+    end
+
+    exitstatus
```
Comment (acammack-r7) [inline lib/msfdb_helpers/pg_ctlcluster.rb:137]:
```suggestion
    output, status = Open3.capture2e(env, cmd)
    if @options[:debug]
      puts "'#{cmd}' returned #{status.exitstatus}"
      puts output
    end

    status.exitstatus
```

File: lib/msfdb_helpers/standalone.rb [hunk 10]
```diff
@@ -0,0 +1,65 @@
+require 'msfdb_helpers/db_interface'
+
+class Standalone < DbInterface
+
+  def initialize(options:, msf_pass:, msftest_pass:, db_conf:)
+    @options = options
+    @msf_pass = msf_pass
+    @msftest_pass = msftest_pass
+    @db_conf = db_conf
+    @conn = PG.connect(host: '127.0.0.1', dbname: 'postgres', port: @options[:db_port], user: @options[:msf_db_admin_username], password: @options[:msf_db_admin_pass])
```
[the remark this answers is not in the document]
Comment (dwelch-r7) [reply lib/msfdb_helpers/standalone.rb:10]: nope, it was just easier to pass the params in like this than smush them together into a connection string

[the remark this answers is not in the document]
Comment (adfoster-r7) [reply lib/msfdb_helpers/standalone.rb:10]: The intent was for the user to provide the connection string directly themselves 👀

File: msfdb [hunk 11]
```diff
@@ -920,6 +799,18 @@ def parse_args(args)
...
 
+    opts.on('--db-admin-pass PASS', "Database admin password (default: #{@options[:msf_db_admin_pass]})") { |p|
+      @options[:msf_db_admin_pass] = p
+    }
+
+    opts.on('--db-admin-username NAME', "Database admin username (default: #{@options[:msf_db_admin_username]}") { |p|
+      @options[:msf_db_admin_username] = p
+    }
+
+    opts.on('--standalone', 'Use a pre-existing database cluster for initialization') {
```
Comment (adfoster-r7) [inline msfdb:810]: Thoughts on consolidating this into just `--connection-string`? That way the user can configure completely arbitrary values, rather than the current hard coded `127.0.0.1` host and `postgres` database name 👀

**Comment by adfoster-r7** [comment]
Maybe it's worth while keeping track of how the database was initialised inside the database.yml, that way we can handle `reinit` gracefully in the future and potentially ask the user what they really mean/intend

File: msfdb [hunk 12]
```diff
@@ -1097,6 +987,19 @@ if $PROGRAM_NAME == __FILE__
...
+  update_db_port
+
+  if @standalone
+    @db_driver = Standalone.new(options: @options, msf_pass: @msf_pass, msftest_pass: @msftest_pass, db_conf: @db_conf)
+  elsif installed?("pg_ctl") && has_requirements(PgCtl.requirements)
+    @db_driver = PgCtl.new(db_path: @db, options: @options, localconf: @localconf, db_conf: @db_conf)
+  elsif installed?("pg_ctlcluster") && has_requirements(PgCtlcluster.requirements)
+    @db_driver = PgCtlcluster.new(db_path: @db, options: @options, localconf: @localconf, db_conf: @db_conf)
+  else
+    abort
```
Comment (adfoster-r7) [inline msfdb:999]: The fall through case here isn't entirely clear, I think before it warned you that you needed particular requirements in order to run msfdb - but now I think it just aborts? 👀

File: msfdb [hunk 13]
```diff
@@ -50,10 +58,13 @@ require 'msfenv'
 @options = {
     component: :all,
     debug: false,
+    msf_db_admin_pass: '',
+    msf_db_admin_username: 'postgres',
```
Comment (adfoster-r7) [inline msfdb:62]: Are these dead now with the connection string functionality?

Comment (dwelch-r7) [reply msfdb:62]: sure are, removing now

File: msfdb [hunk 14]
```diff
@@ -1097,6 +988,20 @@ if $PROGRAM_NAME == __FILE__
...
+  update_db_port
+
+  if @connection_string
+    @db_driver = Standalone.new(options: @options, db_conf: @db_conf, connection_string: @connection_string)
+  elsif installed?('pg_ctl') && has_requirements(PgCtl.requirements)
+    @db_driver = PgCtl.new(db_path: @db, options: @options, localconf: @localconf, db_conf: @db_conf)
+  elsif installed?('pg_ctlcluster') && has_requirements(PgCtlcluster.requirements)
+    @db_driver = PgCtlcluster.new(db_path: @db, options: @options, localconf: @localconf, db_conf: @db_conf)
+  else
+    print_error('You must have pg_ctl or pgctl_cluster on your path, alternatively you can specify a database with `--connection-string`')
```
Comment (adfoster-r7) [inline msfdb:1000]: This seems like a pretty low level message to show to the average user. Is it worth summarising that they most likely need to install postgres?

File: msfdb [hunk 15]
```diff
@@ -920,6 +800,18 @@ def parse_args(args)
...
 
+    opts.on('--db-admin-pass PASS', "Database admin password (default: #{@options[:msf_db_admin_pass]})") { |p|
+      @options[:msf_db_admin_pass] = p
+    }
+
+    opts.on('--db-admin-username NAME', "Database admin username (default: #{@options[:msf_db_admin_username]}") { |p|
+      @options[:msf_db_admin_username] = p
+    }
+
+    opts.on('--connection-string URI', 'Use a pre-existing database cluster for initialization') { |c|
```
Comment (adfoster-r7) [inline msfdb:801]: We should probably add an example that would work here

File: msfdb [hunk 16]
```diff
@@ -920,6 +800,18 @@ def parse_args(args)
...
     }
 
+    opts.on('--db-admin-pass PASS', "Database admin password (default: #{@options[:msf_db_admin_pass]})") { |p|
+      @options[:msf_db_admin_pass] = p
+    }
+
+    opts.on('--db-admin-username NAME', "Database admin username (default: #{@options[:msf_db_admin_username]}") { |p|
+      @options[:msf_db_admin_username] = p
+    }
+
```
Comment (adfoster-r7) [inline msfdb]:
```suggestion
```

✂️

File: msfdb [hunk 17]
```diff
@@ -1097,6 +988,20 @@ if $PROGRAM_NAME == __FILE__
...
+  elsif installed?('pg_ctl') && has_requirements(PgCtl.requirements)
+    @db_driver = PgCtl.new(db_path: @db, options: @options, localconf: @localconf, db_conf: @db_conf)
+  elsif installed?('pg_ctlcluster') && has_requirements(PgCtlcluster.requirements)
+    @db_driver = PgCtlcluster.new(db_path: @db, options: @options, localconf: @localconf, db_conf: @db_conf)
+  else
+    print_error('You must have pg_ctl or pgctl_cluster on your path, alternatively you can specify a database with `--connection-string`')
+    abort
+  end
+
+
```
Comment (adfoster-r7) [inline msfdb:994]:
```suggestion
```

File: lib/msfdb_helpers/pg_ctlcluster.rb [hunk 18]
```diff
@@ -0,0 +1,140 @@
+require 'msfdb_helpers/db_interface'
+
+class PgCtlcluster < DbInterface
```
Comment (adfoster-r7) [inline lib/msfdb_helpers/pg_ctlcluster.rb:3]: Just confirming, should these be in a MsfdbHelpers module/namespace or similar? 👀

File: msfdb (hunk 14 above)
Comment (dwelch-r7) [reply msfdb:990]: what about "You need to have postgres install or specifcy a database with `--connection-string`" ?

File: lib/msfdb_helpers/pg_ctlcluster.rb (hunk 18 above)
Comment (dwelch-r7) [reply lib/msfdb_helpers/pg_ctlcluster.rb:3]: probably should be, will do

**Comment by adfoster-r7** [comment]
Tried running this on Kali:
```
kali@kali:/mnt/hgfs/metasploit-framework$ bundle exec ruby ./msfdb reinit
[?] Would you like to delete your existing data and configurations?: yes
====================================================================
Running the 'reinit' command for the database:
MSF web service is no longer running
Deleting all data at /home/kali/.msf4/db
Creating database at /home/kali/.msf4/db
Starting database at /home/kali/.msf4/db...failed
Creating database users
Traceback (most recent call last):
        10: from ./msfdb:998:in `<main>'
         9: from ./msfdb:998:in `each'
         8: from ./msfdb:1001:in `block in <main>'
         7: from ./msfdb:886:in `invoke_command'
         6: from ./msfdb:303:in `reinit_db'
         5: from ./msfdb:172:in `init_db'
         4: from /mnt/hgfs/metasploit-framework/lib/msfdb_helpers/pg_ctlcluster.rb:43:in `init'
         3: from /mnt/hgfs/metasploit-framework/lib/msfdb_helpers/pg_ctlcluster.rb:135:in `create_db_users'
         2: from /var/lib/gems/2.7.0/gems/pg-1.2.3/lib/pg.rb:58:in `connect'
         1: from /var/lib/gems/2.7.0/gems/pg-1.2.3/lib/pg.rb:58:in `new'
/var/lib/gems/2.7.0/gems/pg-1.2.3/lib/pg.rb:58:in `initialize': FATAL:  password authentication failed for user "msf" (PG::ConnectionBad)
FATAL:  password authentication failed for user "msf"
kali@kali:/mnt/hgfs/metasploit-framework$
```

Didn't look further, just wanted to confirm if it's been tested with Kali yet? 👀

File: lib/msfdb_helpers/standalone.rb [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+    def status
+      raise NotImplementedError
+    end
+
+    def write_db_client_auth_config
+      raise NotImplementedError
+    end
+
+    def self.requirements
+      Array.new
```
Comment (adfoster-r7) [inline lib/msfdb_helpers/standalone.rb:63]: Can we run rubocop on these new files and change`Array.new` to `[]` 👮‍♂️

File: lib/msfdb_helpers/standalone.rb [hunk 20]
```diff
@@ -0,0 +1,67 @@
+require 'msfdb_helpers/db_interface'
+module MsfdbHelpers
+  class Standalone < DbInterface
+
+    def initialize(options:, db_conf:, connection_string:)
+      @options = options
+      @db_conf = db_conf
+      @conn = PG.connect(connection_string)
```
Comment (adfoster-r7) [inline lib/msfdb_helpers/standalone.rb:8]: Verified docker support works; Although I'm think we should clean up this stack trace when the credentials don't work:

```
 bundle exec ruby ./msfdb reinit --connection-string=postgresql://postgres:mysecretpassword@localhost:9000/postgres
Using database port 6000 found in /Users/user/.msf4/database.yml
Traceback (most recent call last):
	5: from ./msfdb:985:in `<main>'
	4: from ./msfdb:985:in `new'
	3: from /Users/user/Documents/code/metasploit-framework/lib/msfdb_helpers/standalone.rb:8:in `initialize'
	2: from /Users/user/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/pg-1.2.3/lib/pg.rb:58:in `connect'
	1: from /Users/user/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/pg-1.2.3/lib/pg.rb:58:in `new'
/Users/adfoster/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/pg-1.2.3/lib/pg.rb:58:in `initialize': could not connect to server: Connection refused (PG::ConnectionBad)
	Is the server running on host "localhost" (127.0.0.1) and accepting
	TCP/IP connections on port 9000?
could not connect to server: Connection refused
	Is the server running on host "localhost" (::1) and accepting
	TCP/IP connections on port 9000?
```

Dumping the stack trace I'm somewhat okay with, but a well placed `print_error('Could not connect to standalone PostgreSQL instance. Ensure that the connection string is valid, and that the database is accessible')`, or similar, could help IMO

File: lib/msfdb_helpers/standalone.rb (hunk 19 above)
Comment (dwelch-r7) [reply lib/msfdb_helpers/standalone.rb:63]: rubocop has already been run but I can change that sure, is that something you expected rubocop to handle?

**Comment by adfoster-r7** [comment]
@dwelch-r7 Good to undraft this? 👀

**Comment by adfoster-r7** [comment]
## Release notes

Updates msfdb to work on additional platforms. Specifically Ubuntu through pg_ctlcluster, as well as an existing or remote databases with the new `--connection-string` option. This option can be used to interact with docker PostgreSQL containers

File: lib/msfdb_helpers/standalone.rb (hunk 20 above)
Comment (adfoster-r7) [reply lib/msfdb_helpers/standalone.rb]: Future me wished past me documented this a bit better; I set up the database with:
```
sudo docker run -p 9000:5432 -e POSTGRES_PASSWORD=mysecretpassword postgres:14.2
```

Will have to get this workflow added to the docs 📚
